src/thread.rs

```diff
@@ -0,0 +1,483 @@
+        assert_ne!(joined_thread_id, self.active_thread, "Bug: trying to join itself");
+        assert!(
+            self.threads
+                .iter()
+                .all(|thread| thread.state != ThreadState::BlockedOnJoin(joined_thread_id)),
+            "Bug: multiple threads try to join the same thread."
+        );
+        if self.threads[joined_thread_id].state != ThreadState::Terminated {
+            // The joined thread is still running, we need to wait for it.
+            self.active_thread_mut().state = ThreadState::BlockedOnJoin(joined_thread_id);
+            trace!(
+                "{:?} blocked on {:?} when trying to join",
+                self.active_thread,
+                joined_thread_id
+            );
+        }
+    }
+
+    /// Set the name of the active thread.
+    fn set_thread_name(&mut self, new_thread_name: Vec<u8>) {
+        self.active_thread_mut().thread_name = Some(new_thread_name);
+    }
+
+    /// Allocate a new blockset id.
+    fn create_blockset(&mut self) -> BlockSetId {
+        self.blockset_counter = self.blockset_counter.checked_add(1).unwrap();
+        self.blockset_counter.into()
+    }
+
+    /// Block the currently active thread and put it into the given blockset.
+    fn block_active_thread(&mut self, set: BlockSetId) {
+        let state = &mut self.active_thread_mut().state;
+        assert_eq!(*state, ThreadState::Enabled);
+        *state = ThreadState::Blocked(set);
+    }
+
+    /// Unblock any one thread from the given blockset if it contains at least
+    /// one. Return the id of the unblocked thread.
+    fn unblock_random_thread(&mut self, set: BlockSetId) -> Option<ThreadId> {
+        for (id, thread) in self.threads.iter_enumerated_mut() {
+            if thread.state == ThreadState::Blocked(set) {
+                trace!("unblocking {:?} in blockset {:?}", id, set);
+                thread.state = ThreadState::Enabled;
+                return Some(id);
+            }
+        }
+        None
+    }
+
+    /// Change the active thread to some enabled thread.
+    fn yield_active_thread(&mut self) {
+        self.yield_active_thread = true;
+    }
+
+    /// Decide which action to take next and on which thread.
+    fn schedule(&mut self) -> InterpResult<'tcx, SchedulingAction> {
+        if self.threads[self.active_thread].check_terminated() {
+            // Check if we need to unblock any threads.
+            for (i, thread) in self.threads.iter_enumerated_mut() {
+                if thread.state == ThreadState::BlockedOnJoin(self.active_thread) {
+                    trace!("unblocking {:?} because {:?} terminated", i, self.active_thread);
+                    thread.state = ThreadState::Enabled;
+                }
+            }
+            return Ok(SchedulingAction::ExecuteDtors);
+        }
+        if self.threads[self.active_thread].state == ThreadState::Enabled
+            && !self.yield_active_thread
```

Comment:
So we run each thread maximally long, never preempting?

I guess we should document the scheduling policy somewhere. This seems like a rather extreme one.

Reply:
> So we run each thread maximally long, never preempting?

Yes. (Unless required.)

> I guess we should document the scheduling policy somewhere.

I have added a comment to the `schedule` function. Do you think this is enough or should I add also somewhere more prominently such as README?

> This seems like a rather extreme one.

As far as I know, stateless model checkers typically use such a scheduler to get the initial trace.

Reply:
Hm, I guess a `schedule` comment is enough.
Longer-term I think it would be nice to use the machine RNG (that was also use for int-ptr-casts) to affect the schedule. Then people could use different seeds to get different schedules, but for a fixed seed everything is still deterministic.